swift/ql/lib/codeql/swift/security/PathInjection.qll

```diff
@@ -0,0 +1,123 @@
+        ";FileManager;true;contentsOfDirectory(at:includingPropertiesForKeys:options:);;;Argument[0];path-injection",
+        ";FileManager;true;contentsOfDirectory(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;enumerator(at:includingPropertiesForKeys:options:errorHandler:);;;Argument[0];path-injection",
+        ";FileManager;true;enumerator(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;subpathsOfDirectory(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;subpaths(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;createDirectory(at:withIntermediateDirectories:attributes:);;;Argument[0];path-injection",
+        ";FileManager;true;createDirectory(atPath:withIntermediateDirectories:attributes:);;;Argument[0];path-injection",
+        ";FileManager;true;createFile(atPath:contents:attributes:);;;Argument[0];path-injection",
+        ";FileManager;true;removeItem(at:);;;Argument[0];path-injection",
+        ";FileManager;true;removeItem(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;trashItem(at:resultingItemURL:);;;Argument[0];path-injection",
+        ";FileManager;true;replaceItem(at:withItemAt:backupItemName:options:resultingItemURL:);;;Argument[0..1];path-injection",
+        ";FileManager;true;replaceItemAt(_:withItemAt:backupItemName:options:);;;Argument[0..1];path-injection",
+        ";FileManager;true;replaceItemAt(_:withItemAt:backupItemName:options:resultingItemURL:);;;Argument[0..1];path-injection",
+        ";FileManager;true;copyItem(at:to:);;;Argument[0..1];path-injection",
+        ";FileManager;true;copyItem(atPath:toPath:);;;Argument[0..1];path-injection",
+        ";FileManager;true;moveItem(at:to:);;;Argument[0..1];path-injection",
+        ";FileManager;true;moveItem(atPath:toPath:);;;Argument[0..1];path-injection",
+        ";FileManager;true;createSymbolicLink(at:withDestinationURL:);;;Argument[0..1];path-injection",
+        ";FileManager;true;createSymbolicLink(atPath:withDestinationPath:);;;Argument[0..1];path-injection",
+        ";FileManager;true;linkItem(at:to:);;;Argument[0..1];path-injection",
+        ";FileManager;true;linkItem(atPath:toPath:);;;Argument[0..1];path-injection",
+        ";FileManager;true;destinationOfSymbolicLink(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;fileExists(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;fileExists(atPath:isDirectory:);;;Argument[0];path-injection",
+        ";FileManager;true;isReadableFile(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;isWritableFile(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;isDeletableFile(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;componentsToDisplay(forPath:);;;Argument[0];path-injection",
+        ";FileManager;true;displayName(atPath:);;;Argument[0];path-injection",
```

Comment:
Yes, I was doubtful about adding certain sinks. My logic ended up being that if the output of these methods was used in a security-relevant check, an attacker could still cause some compromise by altering which file gets used. But in retrospect you're probably right: that is probably very FP-prone, and we should remove some.

I'd remove the following rows, let me know what you think:

```ql
";FileManager;true;isReadableFile(atPath:);;;Argument[0];path-injection",
";FileManager;true;isWritableFile(atPath:);;;Argument[0];path-injection",
";FileManager;true;isDeletableFile(atPath:);;;Argument[0];path-injection",
";FileManager;true;attributesOfItem(atPath:);;;Argument[0];path-injection",
";FileManager;true;attributesOfFileSystem(forPath:);;;Argument[0];path-injection",
";FileManager;true;setAttributes(_:ofItemAtPath:);;;Argument[1];path-injection", // arguably
";FileManager;true;getRelationship(_:ofDirectoryAt:toItemAt:);;;Argument[1..2];path-injection",
";FileManager;true;getRelationship(_:of:in:toItemAt:);;;Argument[3];path-injection",
";FileManager;true;NSHFSTypeOfFile(_:);;;Argument[0];path-injection",

Reply:
`isReadableFile`, `isWritableFile` and the others I'm not so worried about FPs from.  Usually you check a file is readable before you read it, writable before you write it (etc), and I wouldn't want remote data to be controlling that.  I would probably keep them in the results personally, but I'm not too worried either way.

I would keep `setAttributes` - seems like it could be dangerous.

The case of `displayName` and `componentsToDisplay` is special - these are likely to be used to display something to a user, e.g. status or a question, maybe even an error.  It seems to me that we shouldn't flag merely displaying a remote-controlled path unless something is being done with it.

Reply:
>  Usually you check a file is readable before you read it, writable before you write it (etc),

But then the appropriate sinks for read/write would be used anyway, and I'd prefer not to duplicate alerts around the same functionality.

> I would keep setAttributes - seems like it could be dangerous.

👍

Reply:
> But then the appropriate sinks for read/write would be used anyway, and I'd prefer not to duplicate alerts around the same functionality.

Good point.

Reply:
I'm happy with the changes you've made.